crates/red_knot_python_semantic/src/types.rs

```diff
@@ -196,6 +227,9 @@ pub enum Type<'db> {
     LiteralString,
     /// A bytes literal
     BytesLiteral(BytesLiteralType<'db>),
+    /// The set of objects that exist on a given Python version; used for version-conditional definitions.
+    /// Equivalent to Never if the given version is not supported, or 'object' otherwise.
+    ExistsOnVersion(SupportedVersion),
```

Comment:
Hmm, I kind of like this idea. I'm not totally convinced by the performance motivation: we will have _lots_ of intersections due to narrowing constraints in general, and we will need to make our intersection implementation perform as well as we possibly can; I suspect that version/platform narrowing will end up as a very small overall percentage of our intersection types.

But I think it might simplify the implementation to do this, because these types _are_ special: they shouldn't ever exist outside of an intersection, and whenever we encounter them (and haven't already narrowed them away), we should treat them as if they were just the inner type. I think this will be a bit simpler than treating them as `object` and relying on intersection simplification to simplify `foo & object` to `foo`.

Note that we will have `Foo & (PythonVersion[3.11] | PythonVersion[3.12]`, which can't directly be represented this way, but it can distribute to `(Foo & PythonVersion[3.11]) | (Foo & PythonVersion[3.12])` (which we would do anyway, since we normalize to disjunctive normal form), so that should work fine with this representation.